### PR TITLE
Fix PrimaryAllocationIT Race Condition

### LINKED ISSUE #37345: CI: test failure PrimaryAllocationIT.testForceStaleReplicaToBePromotedToPrimaryOnWrongNode
## Example build failure
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+intake/1198/console
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=ubuntu&&virtual/174/console

## Reproduction line 
does not reproduce locally
```
/gradlew :server:integTest -Dtests.seed=ED29DBC5949B19E9 -Dtests.class=org.elasticsearch.cluster.routing.PrimaryAllocationIT -Dtests.method="testForceStaleReplicaToBePromotedToPrimaryOnWrongNode" -Dtests.security.manager=true -Dtests.locale=cs -Dtests.timezone=Europe/Kaliningrad -Dcompiler.java=11 -Druntime.java=8
```

## Example relevant log:
```
10:28:30 ERROR   4.30s J3 | PrimaryAllocationIT.testForceStaleReplicaToBePromotedToPrimaryOnWrongNode <<< FAILURES!
10:28:30    > Throwable #1: java.lang.NullPointerException
10:28:30    > 	at __randomizedtesting.SeedInfo.seed([ED29DBC5949B19E9:D46FFB23117F3451]:0)
10:28:30    > 	at org.elasticsearch.cluster.routing.PrimaryAllocationIT.testForceStaleReplicaToBePromotedToPrimaryOnWrongNode(PrimaryAllocationIT.java:282)
10:28:30    > 	at java.lang.Thread.run(Thread.java:748)
```

## Frequency
up to 8-9 a day.

Possibly related to: #35497

### PR DESCRIPTION
**Pull request by original-brownbear**
* Forcing a stale primary allocation on a green index was tripping an assertion
   * Added a test that this case still errors out correctly
* Made the ability to wipe stopped datanode's data public on the internal test cluster and used it to ensure correct behavior on the fixed test
   * Previously it simply passed because the test finished before the index went green and would NPE when the index was green at the time of the shard store status request, that would then come up empty
* Closes #37345

### CODE REVIEW

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java [hunk 1]
```diff
@@ -1398,8 +1398,7 @@ private void randomlyResetClients() {
         }
     }
 
-    private void wipePendingDataDirectories() {
-        assert Thread.holdsLock(this);
+    public synchronized void wipePendingDataDirectories() {
```
Comment (DaveCTurner) [inline test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: I can't see any _obvious_ problems that arise from exposing this, but the API is kinda complicated so I'd like a second opinion from @ywelsch.

**Review by DaveCTurner — COMMENTED** [review]
Looks ok but I'm always cautious about changes to the `InternalTestCluster` API so I've asked for a second opinion.

File: server/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java [hunk 2]
```diff
@@ -275,7 +274,10 @@ public void testForceStaleReplicaToBePromotedToPrimaryOnWrongNode() throws Excep
                 .put("index.number_of_replicas", 1)).get());
         ensureGreen();
         createStaleReplicaScenario(master);
-        internalCluster().startDataOnlyNodes(2);
+        // Ensure the stopped primary's data is deleted so that it doesn't get picked up by the next datanode we start
+        internalCluster().wipePendingDataDirectories();
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java]: ok as a temporary solution. We definitely need a cleaner way of managing data paths in InternalTestCluster.

**Review by ywelsch — APPROVED** [review]

**Comment by original-brownbear** [comment]
Jenkins run Gradle build tests 2

**Comment by original-brownbear** [comment]
Jenkins run Gradle build tests 2

**Comment by original-brownbear** [comment]
Jenkins run Gradle build tests 1

**Pull request merged**